File: merge_sort/MergeSort.cs
/*
Time Complexity: O(n log(n))
Space Complexity: O(n)
*/

using System;

public class MergeSort
{
    /// <summary>
    /// Merges two unsorted halves of an array into one sorted array
    /// first half: from 'first' to 'mid'
    /// second half: from 'mid' to 'last'
    /// </summary>
    /// <param name="a">The original unsorted array</param>
    /// <param name="first">The start index of the first half of the array a</param>
    /// <param name="mid">The middle index of the two halves of the array</param>
    /// <param name="last">The end index of the second half of the array a</param>
    /// <param name="temp">A temporary array for storing the sorted halves</param>
    public static void Merge(int[] a, int first, int mid, int last, int[] temp)
    {
        int currL = first;  //left half pointer
        int currR = mid;    //right half pointer
        int currT;          //current sorted array pointer

        for (currT = first; currT < last; currT++)
        {
            //if left half is element is smaller or right pointer is out of bounds
            if (currL < mid && (currR >= last || a[currL] < a[currR]))
                temp[currT] = a[currL++];
            else
                temp[currT] = a[currR++];
        }

        for (currT = first; currT < last; currT++)
            a[currT] = temp[currT];
    }

    /// <summary>
    /// Recursive method for sorting an array 'a' between two indexes
    /// first and last by using merge sort
    /// </summary>
    /// <param name="a">The original unsorted array</param>
    /// <param name="first">The start index. The array will be sorted starting from this index</param>
    /// <param name="last">The end index. The array will be sorted untill this index</param>
    /// <param name="temp">A temporary array for storing merge results</param>
    public static void DoMergeSort(int[] a, int first, int last, int[] temp)
    {
        if (last - first > 1)
        {
            int mid = first + (last - first) / 2;   //find the middle
            DoMergeSort(a, first, mid, temp);       //sort left half
            DoMergeSort(a, mid, last, temp);        //sort right half
            Merge(a, first, mid, last, temp);       //merge two sorted halves
        }
    }

    /// <summary>
    /// Merge sort helper method. Calls the recursive Merge sort method
    /// </summary>
    /// <param name="a">The original unsorted array</param>
    public static void DoMergeSort(int[] a)
    {
        if (a == null || a.Length <= 1)
            return;
        DoMergeSort(a, 0, a.Length, new int[a.Length]);
    }

    public static void Main()
    {
        int[] a = new int[] {2, 4, 9, 6, 7, 8};
        Console.WriteLine("Before Merge Sort:");
        Console.WriteLine("[{0}]", string.Join(", ", a));
        DoMergeSort(a);
        Console.WriteLine("After Merge Sort:");
        Console.WriteLine("[{0}]", string.Join(", ", a));
    }
}

File: merge_sort/MergeSort.java
public class MergeSort {

    private static void merge(int[] a, int first, int mid, int last) {
        int l = mid - first + 1;
        int r = last - mid;
        int[] left = new int[l];
        int[] right = new int[r];

        for (int i = 0; i < l; i++) {  //copy left
            left[i] = a[first + i];
        }
        for (int j = 0; j < r; j++) {  //copy right
            right[j] = a[mid + j + 1];
        }
        int i = 0;
        int j = 0;
        int k = first;
        while (i < l && j < r) {  //merging
            if (left[i] <= right[j]) {
                a[k] = left[i];
                i++;
            } else {
                a[k] = right[j];
                j++;
            }
            k++;
        }
        while (i < l) {  //remainig left element
            a[k] = left[i];
            i++;
            k++;
        }
        while (j < r) {  //remainig right element
            a[k] = right[j];
            j++;
            k++;
        }
    }

    private static void mergeSort(int[] a, int first, int last) {
        if (first < last) {
            int mid = (first + last) / 2;  //find the middle
            mergeSort(a, first, mid);  //sort left half
            mergeSort(a, mid + 1, last);  //sort right half
            merge(a, first, mid, last);  //merge above two sorted halves
        }
    }

    public static void main(String[] args) {
        int[] arr = new int[] {2, 4, 9, 6, 7, 8};
        mergeSort(arr, 0, arr.length - 1);
        for (int element : arr) {
            System.out.println(element);
        }
    }
}
